### PR TITLE
Remove listed models from Compute plan event

### PR DESCRIPTION
**Pull request by thibaultrobert**
Companion PR : https://github.com/SubstraFoundation/substra-tests/pull/86 & https://github.com/SubstraFoundation/substra-chaincode/pull/106 & https://github.com/SubstraFoundation/substra/pull/135

### CODE REVIEW

File: backend/substrapp/tasks/tasks.py [hunk 1]
```diff
@@ -897,19 +897,48 @@ def remove_algo_images(algo_hashes):
...
     logger.info(f'Remove local volume of compute plan {compute_plan_id}')
-    remove_local_folder(compute_plan_id)
+
+    if status in ['done', 'failed', 'canceled']:
+        remove_local_folder(compute_plan_id)
 
     # Remove algorithm images
-    remove_algo_images(algo_hashes)
+    if algo_hashes:
+        remove_algo_images(algo_hashes)
```
Comment (thibaultrobert) [inline backend/substrapp/tasks/tasks.py]: We removed the algo images at the end of the compute plan I think.

File: backend/substrapp/serializers/ledger/computeplan/serializer.py [hunk 2]
```diff
@@ -137,6 +138,7 @@ def get_args(self, data):
             'compositeTraintuples': composite_traintuples,
             'aggregatetuples': aggregatetuples,
             'tag': data.get('tag'),
+            'cleanModels': data.get('cleanModels', False),
```
Comment (samlesu) [inline backend/substrapp/serializers/ledger/computeplan/serializer.py]:
```suggestion
            'cleanModels': data.get('clean_models', False),
```

To be consistent with the other endpoints, substra backend should accept snake case fields.

**Comment by Kelvin-M** [comment]
@samlesu The docker/start.py is relatrd to the current PR. In the docker setup, the database used is the same, to the model asset removal remove everything from the db, without checking that the model is, in fact, owned by the node.
We need to split the service for both node to isolate the database.

**Comment by samlesu** [comment]
Got it, thanks for explanation.
If I understand well, when using the docker-compose setup, a single Postgres database instance was used by the 2 nodes?

**Comment by Kelvin-M** [comment]
@samlesu  Yes, so when receiving the event compute plan with the full list of model to delete (for every nodes), each node tries to delete the full list. It will successfully delete the db instance (as the db is shared) but the delete signal inherited will be able to delete model files that the node owned. So we will have a clean db, but not a clean file storage.

File: backend/substrapp/serializers/ledger/computeplan/serializer.py (hunk 2 above)
Comment (Kelvin-M) [reply backend/substrapp/serializers/ledger/computeplan/serializer.py]: Done

File: backend/substrapp/tasks/tasks.py [hunk 3]
```diff
@@ -916,19 +919,43 @@ def remove_algo_images(algo_hashes):
     client = docker.from_env()
     for algo_hash in algo_hashes:
         algo_docker = get_algo_image_name(algo_hash)
-        logger.info(f'Remove docker image {algo_docker}')
-        client.images.remove(algo_docker, force=True)
+
+        try:
+            if client.images.get(algo_docker):
+                client.images.remove(algo_docker, force=True)
+                logger.info(f'Remove docker image {algo_docker}')
```
Comment (AurelienGasser) [inline backend/substrapp/tasks/tasks.py]: swap with the previous line?

File: backend/substrapp/tasks/tasks.py [hunk 4]
```diff
@@ -916,19 +919,43 @@ def remove_algo_images(algo_hashes):
...
+            logger.exception(e)
+
+
+def remove_intermediary_models(model_hashes):
+    from substrapp.models import Model
+
+    models = Model.objects.filter(pkhash__in=model_hashes, validated=True)
+    filtered_model_hashes = [model.pk for model in models]
+
+    models.delete()
```
Comment (AurelienGasser) [inline backend/substrapp/tasks/tasks.py]: aren't the models also stored on disk?

Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py]: Yes. Models are stored on the disk :-)
But there is a model signal that delete it when the instance is deleted !

File: backend/substrapp/tasks/tasks.py (hunk 3 above)
Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py]: Yes ! Nice catch

Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py]: Done

File: backend/substrapp/tasks/tasks.py (hunk 4 above)
Comment (AurelienGasser) [reply backend/substrapp/tasks/tasks.py]: Would you mind pointing me to the piece of code that does that?

Also: have you checked the files actually get deleted on disk?

Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py]: You can find the code here : https://github.com/SubstraFoundation/substra-backend/blob/master/backend/substrapp/signals/model/post_delete.py
Linked here : https://github.com/SubstraFoundation/substra-backend/blob/master/backend/substrapp/apps.py#L26
Yes, I've checked that models are removed from db and deleted on disk

Comment (AurelienGasser) [reply backend/substrapp/tasks/tasks.py]: Thnks for the links!

**Pull request merged**